// @umijs/plugin-blocks/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _umi() {
  const data = require("umi");

  _umi = function _umi() {
    return data;
  };

  return data;
}

function _blockSdk() {
  const data = require("@umijs/block-sdk");

  _blockSdk = function _blockSdk() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const signale = _umi().utils.signale,
      createDebug = _umi().utils.createDebug,
      chalk = _umi().utils.chalk;

const debug = createDebug('plugin-blocks');

var _default = api => {
  const userConfig = api.userConfig;
  const blockConfig = (userConfig === null || userConfig === void 0 ? void 0 : userConfig.block) || {};
  api.describe({
    key: 'block',
    config: {
      schema(joi) {
        return joi.number();
      }

    }
  });
  debug(`blockConfig ${blockConfig}`);

  function block() {
    return _block.apply(this, arguments);
  }

  function _block() {
    _block = _asyncToGenerator(function* (args = {}, opts = {}) {
      let retCtx;

      switch (args._[0]) {
        case 'clear':
          yield (0, _blockSdk().clearGitCache)({
            dryRun: args.dryRun
          });
          break;

        case 'add':
          retCtx = yield (0, _blockSdk().addBlock)(_objectSpread(_objectSpread({}, args), {}, {
            url: args._[1]
          }), opts, api);
          break;

        case 'list':
          retCtx = yield (0, _blockSdk().getDefaultBlockList)(args, blockConfig, api);
          break;

        default:
          throw new Error(`Please run ${chalk.cyan.underline('umi help block')} to checkout the usage`);
      }

      return retCtx; // return for test
    });
    return _block.apply(this, arguments);
  }

  const details = `

Commands:

  ${chalk.cyan(`add `)}     add a block to your project
  ${chalk.cyan(`list`)}     list all blocks
  ${chalk.cyan(`clear`)}    clear all git cache


Options for the ${chalk.cyan(`add`)} command:

  ${chalk.green(`--path              `)} the file path, default the name in package.json
  ${chalk.green(`--route-path        `)} the route path, default the name in package.json
  ${chalk.green(`--branch            `)} git branch
  ${chalk.green(`--npm-client        `)} the npm client, default npm or yarn (if has yarn.lock)
  ${chalk.green(`--skip-dependencies `)} don't install dependencies
  ${chalk.green(`--skip-modify-routes`)} don't modify the routes
  ${chalk.green(`--dry-run           `)} for test, don't install dependencies and download
  ${chalk.green(`--page              `)} add the block to a independent directory as a page
  ${chalk.green(`--layout            `)} add as a layout block (add route with empty children)
  ${chalk.green(`--js                `)} If the block is typescript, convert to js
  ${chalk.green(`--registry          `)} set up npm installation using the registry
  ${chalk.green(`--uni18n            `)} remove umi-plugin-locale formatMessage
  ${chalk.green(`--closeFastGithub   `)} If using custom block repository, please set it to true

Examples:

  ${chalk.gray(`# Add block`)}
  umi block add demo
  umi block add ant-design-pro/Monitor

  ${chalk.gray(`# Add block with full url`)}
  umi block add https://github.com/umijs/umi-blocks/tree/master/blocks/demo

  ${chalk.gray(`# Add block with specified route path`)}
  umi block add demo --path /foo/bar

  ${chalk.gray(`# List all blocks`)}
  umi block list
  `.trim();
  api.registerCommand({
    name: 'block',

    fn({
      args
    }) {
      return _asyncToGenerator(function* () {
        if (!args._[0]) {
          // TODO: use plugin register args
          console.log(details.split('\n').map(line => `  ${line}`).join('\n'));
          return;
        } // return only for test


        try {
          yield block(args);
        } catch (e) {
          signale.error(e);
        }
      })();
    } // description: 'block related commands, e.g. add, list',
    // usage: `umi block <command>`,
    // details,


  });
};

exports.default = _default;
