// @umijs/preset-ui/lib/plugins/configuration/ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@ant-design/icons");

var _antd = require("antd");

var _index = _interopRequireDefault(require("./ui/index"));

var _zhCN = _interopRequireDefault(require("./locales/zh-CN"));

var _enUS = _interopRequireDefault(require("./locales/en-US"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ConfigAction = function ConfigAction(_ref) {
  var api = _ref.api;
  var intl = api.intl;
  var FormattedMessage = intl.FormattedMessage;
  return /*#__PURE__*/_react.default.createElement(_antd.Button, {
    onClick: function onClick() {
      api.launchEditor({
        type: 'config'
      });
    },
    size: api.mini ? 'small' : 'default',
    type: "default"
  }, /*#__PURE__*/_react.default.createElement(FormattedMessage, {
    id: "org.umi.ui.configuration.actions.open.config"
  }));
};

var _default = function _default(api) {
  api.addLocales({
    'zh-CN': _zhCN.default,
    'en-US': _enUS.default
  });
  var FormattedMessage = api.intl.FormattedMessage;
  api.addDashboard({
    key: 'org.umi.dashboard.card.config',
    // default enable
    enable: false,
    title: /*#__PURE__*/_react.default.createElement(FormattedMessage, {
      id: "org.umi.ui.configuration.panel"
    }),
    description: /*#__PURE__*/_react.default.createElement(FormattedMessage, {
      id: "org.umi.ui.configuration.panel.description"
    }),
    icon: /*#__PURE__*/_react.default.createElement(_icons.ControlFilled, null),
    content: [/*#__PURE__*/_react.default.createElement("a", {
      onClick: function onClick() {
        return api.redirect('/configuration');
      }
    }, /*#__PURE__*/_react.default.createElement(FormattedMessage, {
      id: "org.umi.ui.configuration.project.config.title"
    }))]
  });
  api.addPanel({
    title: 'org.umi.ui.configuration.panel',
    actions: [/*#__PURE__*/_react.default.createElement(ConfigAction, {
      api: api
    })],
    path: '/configuration',
    icon: /*#__PURE__*/_react.default.createElement(_icons.ControlFilled, null),
    component: function component() {
      return /*#__PURE__*/_react.default.createElement(_index.default, {
        api: api
      });
    }
  });
};

exports.default = _default;
