// @umijs/plugin-ui-tasks/lib/server/core/enums/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NpmClient = exports.LogType = exports.TaskEventType = exports.TaskType = exports.TaskState = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let TaskState;
exports.TaskState = TaskState;

(function (TaskState) {
  TaskState["INIT"] = "INIT";
  TaskState["ING"] = "ING";
  TaskState["SUCCESS"] = "SUCCESS";
  TaskState["FAIL"] = "FAIL";
})(TaskState || (exports.TaskState = TaskState = {}));

let TaskType;
exports.TaskType = TaskType;

(function (TaskType) {
  TaskType["BUILD"] = "BUILD";
  TaskType["DEV"] = "DEV";
  TaskType["TEST"] = "TEST";
  TaskType["LINT"] = "LINT";
  TaskType["INSTALL"] = "INSTALL";
})(TaskType || (exports.TaskType = TaskType = {}));

let TaskEventType;
exports.TaskEventType = TaskEventType;

(function (TaskEventType) {
  TaskEventType["STD_OUT_DATA"] = "STD_OUT_DATA";
  TaskEventType["STD_ERR_DATA"] = "STD_ERR_DATA";
  TaskEventType["STATE_EVENT"] = "STATE_EVENT";
})(TaskEventType || (exports.TaskEventType = TaskEventType = {}));

let LogType;
exports.LogType = LogType;

(function (LogType) {
  LogType["STD_OUT"] = "STD_OUT";
  LogType["STD_ERR"] = "STD_ERR";
})(LogType || (exports.LogType = LogType = {}));

let NpmClient;
exports.NpmClient = NpmClient;

(function (NpmClient) {
  NpmClient["tnpm"] = "tnpm";
  NpmClient["cnpm"] = "cnpm";
  NpmClient["npm"] = "npm";
  NpmClient["ayarn"] = "ayarn";
  NpmClient["tyarn"] = "tyarn";
  NpmClient["yarn"] = "yarn";
  NpmClient["pnpm"] = "pnpm";
})(NpmClient || (exports.NpmClient = NpmClient = {}));
