// @umijs/plugin-ui-tasks/lib/server/core/Tasks/Build.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BuildTask = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

var _Base = require("./Base");

var _enums = require("../enums");

var _util = require("../../util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const STATS_FILE_NAME = 'webpack-stats-build.json';

class BuildTask extends _Base.BaseTask {
  // 项目的 db 路径
  constructor(opts) {
    super(opts);
    this.dbPath = void 0;
    this.webpackStats = void 0;
    this.type = _enums.TaskType.BUILD;
  }

  run(args, env = {}) {
    var _superprop_getRun = () => super.run,
        _this = this;

    return _asyncToGenerator(function* () {
      _this.webpackStats = null;
      yield _superprop_getRun().call(_this);

      const _this$getScript = _this.getScript(),
            script = _this$getScript.script,
            scriptEnvs = _this$getScript.envs;

      const analyzeEnv = yield (0, _util.getBuildAnalyzeEnv)({
        analyze: args.analyze,
        dbPath: args.dbPath,
        fileName: STATS_FILE_NAME
      });
      _this.proc = yield (0, _util.runCommand)(script, {
        cwd: _this.cwd,
        env: _objectSpread(_objectSpread(_objectSpread({}, env), analyzeEnv), scriptEnvs)
      });

      _this.handleChildProcess(_this.proc); // 进度条更新


      _this.proc.on('message', msg => {
        if (_this.state !== _enums.TaskState.ING) {
          return;
        }

        const type = msg.type;

        if (type === 'STARTING') {
          _this.updateProgress(msg);
        }
      });
    })();
  }

  getDetail(dbPath) {
    var _superprop_getGetDetail = () => super.getDetail,
        _this2 = this;

    return _asyncToGenerator(function* () {
      if (dbPath) {
        _this2.dbPath = dbPath;
      }

      return _objectSpread(_objectSpread({}, yield _superprop_getGetDetail().call(_this2)), {}, {
        progress: _this2.progress,
        stats: yield _this2.getStats()
      });
    })();
  }

  getStats() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      if (_this3.state !== _enums.TaskState.SUCCESS || !_this3.dbPath) {
        return null;
      }

      if (_this3.webpackStats) {
        return _this3.webpackStats;
      }

      _this3.webpackStats = (0, _util.getChartData)((0, _path().join)(_this3.dbPath, STATS_FILE_NAME));
      return _this3.webpackStats;
    })();
  }

  getScript() {
    const _parseScripts = (0, _util.parseScripts)({
      pkgPath: this.pkgPath,
      key: 'build'
    }),
          succes = _parseScripts.succes,
          exist = _parseScripts.exist,
          errMsg = _parseScripts.errMsg,
          envs = _parseScripts.envs,
          bin = _parseScripts.bin,
          args = _parseScripts.args;

    if (!exist) {
      return {
        script: this.isBigfishProject ? 'bigfish build' : 'umi build',
        envs: []
      };
    }

    if (!succes) {
      this.error(errMsg);
    }

    return {
      script: `${bin} ${args.join(' ')}`,
      envs
    };
  }

}

exports.BuildTask = BuildTask;
