// @umijs/plugin-test/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _umi() {
  const data = require("umi");

  _umi = function _umi() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _default = api => {
  api.registerCommand({
    name: 'test',
    description: 'test with jest',
    details: `
$ umi-test

# watch mode
$ umi-test -w
$ umi-test --watch

# collect coverage
$ umi-test --coverage

# print debug info
$ umi-test --debug

# test specified package for lerna package
$ umi-test --package name

# don't do e2e test
$ umi-test --no-e2e
    `,

    fn() {
      return _asyncToGenerator(function* () {
        const args = _umi().utils.yParser(process.argv.slice(3), {
          alias: {
            watch: ['w'],
            version: ['v']
          },
          boolean: ['coverage', 'watch', 'version', 'debug', 'e2e'],
          default: {
            e2e: true
          }
        });

        require('@umijs/test').default(args).catch(e => {
          console.error(_umi().utils.chalk.red(e));
          process.exit(1);
        });
      })();
    }

  });
};

exports.default = _default;
