// @formatjs/intl-utils/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var diff_1 = require("./diff");
exports.selectUnit = diff_1.selectUnit;
var polyfill_utils_1 = require("./polyfill-utils");
exports.defaultNumberOption = polyfill_utils_1.defaultNumberOption;
exports.getAliasesByLang = polyfill_utils_1.getAliasesByLang;
exports.getInternalSlot = polyfill_utils_1.getInternalSlot;
exports.getMultiInternalSlots = polyfill_utils_1.getMultiInternalSlots;
exports.getNumberOption = polyfill_utils_1.getNumberOption;
exports.getOption = polyfill_utils_1.getOption;
exports.getParentLocalesByLang = polyfill_utils_1.getParentLocalesByLang;
exports.isLiteralPart = polyfill_utils_1.isLiteralPart;
exports.partitionPattern = polyfill_utils_1.partitionPattern;
exports.setInternalSlot = polyfill_utils_1.setInternalSlot;
exports.setMultiInternalSlots = polyfill_utils_1.setMultiInternalSlots;
exports.setNumberFormatDigitOptions = polyfill_utils_1.setNumberFormatDigitOptions;
exports.toObject = polyfill_utils_1.toObject;
exports.objectIs = polyfill_utils_1.objectIs;
exports.isWellFormedCurrencyCode = polyfill_utils_1.isWellFormedCurrencyCode;
exports.toString = polyfill_utils_1.toString;
var resolve_locale_1 = require("./resolve-locale");
exports.createResolveLocale = resolve_locale_1.createResolveLocale;
exports.getLocaleHierarchy = resolve_locale_1.getLocaleHierarchy;
exports.supportedLocales = resolve_locale_1.supportedLocales;
exports.unpackData = resolve_locale_1.unpackData;
exports.isMissingLocaleDataError = resolve_locale_1.isMissingLocaleDataError;
__export(require("./units"));
__export(require("./number-types"));
var get_canonical_locales_1 = require("./get-canonical-locales");
exports.getCanonicalLocales = get_canonical_locales_1.getCanonicalLocales;
var invariant_1 = require("./invariant");
exports.invariant = invariant_1.invariant;
