// umi-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "findJS", {
  enumerable: true,
  get: function get() {
    return _findJS.default;
  }
});
Object.defineProperty(exports, "compatDirname", {
  enumerable: true,
  get: function get() {
    return _compatDirname.default;
  }
});
Object.defineProperty(exports, "winPath", {
  enumerable: true,
  get: function get() {
    return _winPath.default;
  }
});
Object.defineProperty(exports, "optsToArray", {
  enumerable: true,
  get: function get() {
    return _optsToArray.default;
  }
});
Object.defineProperty(exports, "chunkName", {
  enumerable: true,
  get: function get() {
    return _chunkName.default;
  }
});
Object.defineProperty(exports, "endWithSlash", {
  enumerable: true,
  get: function get() {
    return _endWithSlash.default;
  }
});
Object.defineProperty(exports, "isUrl", {
  enumerable: true,
  get: function get() {
    return _isUrl.default;
  }
});
Object.defineProperty(exports, "deprecate", {
  enumerable: true,
  get: function get() {
    return _deprecate.default;
  }
});
Object.defineProperty(exports, "winEOL", {
  enumerable: true,
  get: function get() {
    return _winEOL.default;
  }
});
Object.defineProperty(exports, "prettierFile", {
  enumerable: true,
  get: function get() {
    return _prettierFile.default;
  }
});
Object.defineProperty(exports, "getFastGithub", {
  enumerable: true,
  get: function get() {
    return _getFastGithub.default;
  }
});
Object.defineProperty(exports, "loadDotEnv", {
  enumerable: true,
  get: function get() {
    return _loadDotEnv.default;
  }
});

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

var _findJS = _interopRequireDefault(require("./findJS"));

var _compatDirname = _interopRequireDefault(require("./compatDirname"));

var _winPath = _interopRequireDefault(require("./winPath"));

var _optsToArray = _interopRequireDefault(require("./optsToArray"));

var _chunkName = _interopRequireDefault(require("./chunkName"));

var _endWithSlash = _interopRequireDefault(require("./endWithSlash"));

var _isUrl = _interopRequireDefault(require("./isUrl"));

var _deprecate = _interopRequireDefault(require("./deprecate"));

var _winEOL = _interopRequireDefault(require("./winEOL"));

var _prettierFile = _interopRequireDefault(require("./prettierFile"));

var _getFastGithub = _interopRequireDefault(require("./getFastGithub"));

var _loadDotEnv = _interopRequireDefault(require("./loadDotEnv"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
