// @umijs/plugin-layout/lib/layout/renderRightContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = renderRightContent;

require("antd/es/dropdown/style");

var _dropdown = _interopRequireDefault(require("antd/es/dropdown"));

require("antd/es/spin/style");

var _spin = _interopRequireDefault(require("antd/es/spin"));

require("antd/es/avatar/style");

var _avatar = _interopRequireDefault(require("antd/es/avatar"));

require("antd/es/menu/style");

var _menu = _interopRequireDefault(require("antd/es/menu"));

var _react = _interopRequireDefault(require("react"));

var _umi = require("umi");

var _icons = require("@ant-design/icons");

require("./style.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function renderRightContent(runtimeLayout, loading, initialState, setInitialState) {
  if (runtimeLayout.rightRender) {
    return runtimeLayout.rightRender(initialState, setInitialState, runtimeLayout);
  }

  var menu = /*#__PURE__*/_react.default.createElement(_menu.default, {
    className: "umi-plugin-layout-menu"
  }, /*#__PURE__*/_react.default.createElement(_menu.default.Item, {
    key: "logout",
    onClick: function onClick() {
      return runtimeLayout.logout && (runtimeLayout === null || runtimeLayout === void 0 ? void 0 : runtimeLayout.logout(initialState));
    }
  }, /*#__PURE__*/_react.default.createElement(_icons.LogoutOutlined, null), "\u9000\u51FA\u767B\u5F55"));

  var avatar = /*#__PURE__*/_react.default.createElement("span", {
    className: "umi-plugin-layout-action"
  }, /*#__PURE__*/_react.default.createElement(_avatar.default, {
    size: "small",
    className: "umi-plugin-layout-avatar",
    src: initialState && initialState.avatar || 'https://gw.alipayobjects.com/zos/antfincdn/XAosXuNZyF/BiazfanxmamNRoxxVxka.png',
    alt: "avatar"
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: "umi-plugin-layout-name"
  }, initialState && initialState.name));

  if (loading) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "umi-plugin-layout-right"
    }, /*#__PURE__*/_react.default.createElement(_spin.default, {
      size: "small",
      style: {
        marginLeft: 8,
        marginRight: 8
      }
    }));
  }

  return initialState && /*#__PURE__*/_react.default.createElement("div", {
    className: "umi-plugin-layout-right"
  }, runtimeLayout.logout ? /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    overlay: menu,
    overlayClassName: "umi-plugin-layout-container"
  }, avatar) : avatar, _umi.SelectLang && /*#__PURE__*/_react.default.createElement(_umi.SelectLang, null));
}
