// @umijs/plugin-layout/lib/component/Exception/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WithExceptionOpChildren = exports.Exception500 = exports.Exception403 = exports.Exception404 = void 0;

require("antd/es/result/style");

var _result = _interopRequireDefault(require("antd/es/result"));

require("antd/es/button/style");

var _button = _interopRequireDefault(require("antd/es/button"));

var _react = _interopRequireDefault(require("react"));

var _umi = require("umi");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function backToHome() {
  _umi.history.push('/');
}

var Exception404 = function Exception404() {
  return /*#__PURE__*/_react.default.createElement(_result.default, {
    status: "404",
    title: "404",
    subTitle: "\u62B1\u6B49\uFF0C\u4F60\u8BBF\u95EE\u7684\u9875\u9762\u4E0D\u5B58\u5728",
    extra: /*#__PURE__*/_react.default.createElement(_button.default, {
      type: "primary",
      onClick: backToHome
    }, "\u8FD4\u56DE\u9996\u9875")
  });
};

exports.Exception404 = Exception404;

var Exception500 = function Exception500() {
  return /*#__PURE__*/_react.default.createElement(_result.default, {
    status: "500",
    title: "500",
    subTitle: "\u62B1\u6B49\uFF0C\u670D\u52A1\u5668\u51FA\u9519\u4E86",
    extra: /*#__PURE__*/_react.default.createElement(_button.default, {
      type: "primary",
      onClick: backToHome
    }, "\u8FD4\u56DE\u9996\u9875")
  });
};

exports.Exception500 = Exception500;

var Exception403 = function Exception403() {
  return /*#__PURE__*/_react.default.createElement(_result.default, {
    status: "403",
    title: "403",
    subTitle: "\u62B1\u6B49\uFF0C\u4F60\u65E0\u6743\u8BBF\u95EE\u8BE5\u9875\u9762",
    extra: /*#__PURE__*/_react.default.createElement(_button.default, {
      type: "primary",
      onClick: backToHome
    }, "\u8FD4\u56DE\u9996\u9875")
  });
};
/**
 * 异常路由处理组件
 * - 无权限
 * - 404
 */


exports.Exception403 = Exception403;

var WithExceptionOpChildren = function WithExceptionOpChildren(props) {
  var children = props.children,
      currentPathConfig = props.currentPathConfig; // 404 现在应该很少会发生

  if (!currentPathConfig) {
    return /*#__PURE__*/_react.default.createElement(Exception404, null);
  }
  /**
   * 这里是没有权限的意思
   */


  if (currentPathConfig.unAccessible || currentPathConfig.unaccessible) {
    return /*#__PURE__*/_react.default.createElement(Exception403, null);
  }

  return children;
};

exports.WithExceptionOpChildren = WithExceptionOpChildren;
