// @umijs/preset-ui/lib/plugins/dashboard/ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@ant-design/icons");

var _index = _interopRequireDefault(require("./ui/index"));

var _action = _interopRequireDefault(require("./ui/action"));

var _layout = _interopRequireDefault(require("./ui/layout"));

var _dailyReportTitle = _interopRequireDefault(require("./ui/plugins/dailyReportTitle"));

var _dailyReport = _interopRequireDefault(require("./ui/plugins/dailyReport"));

var _dailyReportHeader = _interopRequireDefault(require("./ui/plugins/dailyReportHeader"));

var _zhCN = _interopRequireDefault(require("./locales/zh-CN"));

var _enUS = _interopRequireDefault(require("./locales/en-US"));

var _indexModule = _interopRequireDefault(require("./ui/index.module.less"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = function _default(api) {
  api.addLocales({
    'zh-CN': _zhCN.default,
    'en-US': _enUS.default
  });
  var FormattedMessage = api.intl.FormattedMessage;
  api.addDashboard({
    key: 'org.umi.dashboard.card.zaobao',
    title: /*#__PURE__*/_react.default.createElement(_dailyReportTitle.default, null),
    description: /*#__PURE__*/_react.default.createElement(FormattedMessage, {
      id: "org.umi.ui.dashboard.card.zaobao.description"
    }),
    icon: 'https://img.alicdn.com/tfs/TB1JJ12nbr1gK0jSZFDXXb9yVXa-225-225.png',
    right: /*#__PURE__*/_react.default.createElement(_dailyReportHeader.default, null),
    colClassName: _indexModule.default['zaobao-col'],
    content: /*#__PURE__*/_react.default.createElement(_dailyReport.default, null)
  });
  api.addPanel({
    title: 'org.umi.ui.dashboard.panel',
    path: '/dashboard',
    actions: [/*#__PURE__*/_react.default.createElement(_layout.default, {
      api: api
    }, /*#__PURE__*/_react.default.createElement(_action.default, null))],
    icon: /*#__PURE__*/_react.default.createElement(_icons.DashboardFilled, null),
    component: function component() {
      return /*#__PURE__*/_react.default.createElement(_layout.default, {
        api: api
      }, /*#__PURE__*/_react.default.createElement(_index.default, null));
    }
  });
};

exports.default = _default;
