// @umijs/plugin-ui-blocks/lib/ui/server/core/tasks/filesTasks/install.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _blockSdk() {
  const data = require("@umijs/block-sdk");

  _blockSdk = function _blockSdk() {
    return data;
  };

  return data;
}

function _umi() {
  const data = require("umi");

  _umi = function _umi() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const createDebug = _umi().utils.createDebug;

const debug = createDebug('umi:umiui:UmiUI:block:filetasks');

const install = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator(function* (ctx, args) {
    const logger = ctx.logger,
          execa = ctx.execa,
          api = ctx.api;
    const registry = ctx.stages.registry;
    yield (0, _blockSdk().installFilesDependencies)({
      npmClient: args.npmClient,
      registry,
      applyPlugins: api.applyPlugins,
      ApplyPluginsType: api.ApplyPluginsType,
      paths: api.paths,
      debug,
      dryRun: args.dryRun,
      spinner: logger,
      skipDependencies: args.skipDependencies,
      execa
    }, ctx.stages.blockCtx);
  });

  return function install(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var _default = install;
exports.default = _default;
