// @umijs/plugin-ui-blocks/lib/ui/server/core/tasks/filesTasks/parse_url.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function _assert() {
    return data;
  };

  return data;
}

function _umi() {
  const data = require("umi");

  _umi = function _umi() {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function _chalk() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _fs() {
  const data = require("fs");

  _fs = function _fs() {
    return data;
  };

  return data;
}

function _getnpmregistry() {
  const data = _interopRequireDefault(require("getnpmregistry"));

  _getnpmregistry = function _getnpmregistry() {
    return data;
  };

  return data;
}

var _helper = require("../helper");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const createDebug = _umi().utils.createDebug,
      winPath = _umi().utils.winPath;

const debug = createDebug('umi:umiui:UmiUI:block:filetasks');
/**
 * 解析 url，
 * @param ctx
 * @param args
 */

var _default = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator(function* (ctx, args) {
    const files = args.files;
    debug('files parse args', args); // ctx.logger.setId(url); // 设置这次 flow 的 log trace id

    ctx.result.files = files; // 记录当前的 url

    (0, _assert().default)(files, `run ${_chalk().default.cyan.underline('umi help block')} to checkout the usage`);
    const _ctx$api = ctx.api,
          config = _ctx$api.config,
          cwd = _ctx$api.cwd;
    const blockConfig = config.block || {};
    const useYarn = (0, _fs().existsSync)((0, _path().join)(cwd, 'yarn.lock'));
    const defaultNpmClient = blockConfig.npmClient || (useYarn ? 'yarn' : 'npm');
    const registryUrl = yield (0, _getnpmregistry().default)(); // setup route path

    const path = args.path;
    const filePath = winPath(path);

    const blockCtx = _objectSpread(_objectSpread({}, args), {}, {
      npmClient: args.npmClient || defaultNpmClient,
      routePath: (0, _helper.addPrefix)(args.routePath || filePath),
      filePath,
      pkg: {
        dependencies: args.dependencies || {},
        peerDependencies: args.peerDependencies || {},
        devDependencies: args.devDependencies || {}
      }
    });

    ctx.stages.blockCtx = blockCtx;
    ctx.stages.registry = args.registry || registryUrl;
  });

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

exports.default = _default;
