// @umijs/ui/lib/Actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OpenProjectAction = exports.OpenConfigFileAction = exports.ReInstallDependencyAction = exports.InstallDependencyAction = exports.BackToHomeAction = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BackToHomeAction = {
  title: {
    "zh-CN": "返回主页",
    "en-US": "Back To Home"
  },
  browserHandler: "BACK_TO_HOME"
};
exports.BackToHomeAction = BackToHomeAction;
const InstallDependencyAction = {
  title: {
    "zh-CN": "安装依赖",
    "en-US": "Install Dependencies"
  },
  handler: {
    type: "@@actions/installDependencies",
    payload: {
      npmClient: true,
      projectPath: true
    }
  }
};
exports.InstallDependencyAction = InstallDependencyAction;
const ReInstallDependencyAction = {
  title: {
    "zh-CN": "重新安装依赖",
    "en-US": "Reinstall Dependencies"
  },
  handler: {
    type: "@@actions/reInstallDependencies",
    payload: {
      npmClient: true,
      projectPath: true
    }
  }
};
exports.ReInstallDependencyAction = ReInstallDependencyAction;
const OpenConfigFileAction = {
  title: {
    "zh-CN": "打开配置文件11",
    "en-US": "Open Configuration File"
  },
  handler: {
    type: "@@actions/openConfigFile",
    payload: {
      projectPath: true
    }
  }
};
exports.OpenConfigFileAction = OpenConfigFileAction;
const OpenProjectAction = {
  title: {
    "zh-CN": "在编辑器中打开项目",
    "en-US": "Open Project In Editor"
  },
  handler: {
    type: "@@actions/openProjectInEditor",
    payload: {
      projectPath: true
    }
  }
};
exports.OpenProjectAction = OpenProjectAction;
