// @umijs/utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  address: true,
  chalk: true,
  spawn: true,
  chokidar: true,
  clipboardy: true,
  createDebug: true,
  deepmerge: true,
  execa: true,
  lodash: true,
  glob: true,
  portfinder: true,
  got: true,
  resolve: true,
  yargs: true,
  mkdirp: true,
  pkgUp: true,
  Mustache: true,
  signale: true,
  rimraf: true,
  yParser: true,
  t: true,
  parser: true,
  traverse: true,
  semver: true,
  cheerio: true,
  compatESModuleRequire: true,
  mergeConfig: true,
  isLernaPackage: true,
  getFile: true,
  winPath: true,
  winEOL: true,
  isWindows: true,
  parseRequireDeps: true,
  BabelRegister: true,
  Generator: true,
  randomColor: true,
  delay: true,
  cleanRequireCache: true
};
Object.defineProperty(exports, "address", {
  enumerable: true,
  get: function get() {
    return _address().default;
  }
});
Object.defineProperty(exports, "chalk", {
  enumerable: true,
  get: function get() {
    return _chalk().default;
  }
});
Object.defineProperty(exports, "spawn", {
  enumerable: true,
  get: function get() {
    return _crossSpawn().default;
  }
});
Object.defineProperty(exports, "clipboardy", {
  enumerable: true,
  get: function get() {
    return _clipboardy().default;
  }
});
Object.defineProperty(exports, "createDebug", {
  enumerable: true,
  get: function get() {
    return _debug().default;
  }
});
Object.defineProperty(exports, "deepmerge", {
  enumerable: true,
  get: function get() {
    return _deepmerge().default;
  }
});
Object.defineProperty(exports, "execa", {
  enumerable: true,
  get: function get() {
    return _execa().default;
  }
});
Object.defineProperty(exports, "lodash", {
  enumerable: true,
  get: function get() {
    return _lodash().default;
  }
});
Object.defineProperty(exports, "glob", {
  enumerable: true,
  get: function get() {
    return _glob().default;
  }
});
Object.defineProperty(exports, "portfinder", {
  enumerable: true,
  get: function get() {
    return _portfinder().default;
  }
});
Object.defineProperty(exports, "got", {
  enumerable: true,
  get: function get() {
    return _got().default;
  }
});
Object.defineProperty(exports, "resolve", {
  enumerable: true,
  get: function get() {
    return _resolve().default;
  }
});
Object.defineProperty(exports, "yargs", {
  enumerable: true,
  get: function get() {
    return _yargs().default;
  }
});
Object.defineProperty(exports, "mkdirp", {
  enumerable: true,
  get: function get() {
    return _mkdirp().default;
  }
});
Object.defineProperty(exports, "pkgUp", {
  enumerable: true,
  get: function get() {
    return _pkgUp().default;
  }
});
Object.defineProperty(exports, "Mustache", {
  enumerable: true,
  get: function get() {
    return _mustache().default;
  }
});
Object.defineProperty(exports, "signale", {
  enumerable: true,
  get: function get() {
    return _signale().default;
  }
});
Object.defineProperty(exports, "rimraf", {
  enumerable: true,
  get: function get() {
    return _rimraf().default;
  }
});
Object.defineProperty(exports, "yParser", {
  enumerable: true,
  get: function get() {
    return _yargsParser().default;
  }
});
Object.defineProperty(exports, "semver", {
  enumerable: true,
  get: function get() {
    return _semver().default;
  }
});
Object.defineProperty(exports, "cheerio", {
  enumerable: true,
  get: function get() {
    return _cheerio.default;
  }
});
Object.defineProperty(exports, "compatESModuleRequire", {
  enumerable: true,
  get: function get() {
    return _compatESModuleRequire.default;
  }
});
Object.defineProperty(exports, "mergeConfig", {
  enumerable: true,
  get: function get() {
    return _mergeConfig.default;
  }
});
Object.defineProperty(exports, "isLernaPackage", {
  enumerable: true,
  get: function get() {
    return _isLernaPackage.default;
  }
});
Object.defineProperty(exports, "getFile", {
  enumerable: true,
  get: function get() {
    return _getFile.default;
  }
});
Object.defineProperty(exports, "winPath", {
  enumerable: true,
  get: function get() {
    return _winPath.default;
  }
});
Object.defineProperty(exports, "winEOL", {
  enumerable: true,
  get: function get() {
    return _winEOL.default;
  }
});
Object.defineProperty(exports, "isWindows", {
  enumerable: true,
  get: function get() {
    return _winEOL.isWindows;
  }
});
Object.defineProperty(exports, "parseRequireDeps", {
  enumerable: true,
  get: function get() {
    return _parseRequireDeps.default;
  }
});
Object.defineProperty(exports, "BabelRegister", {
  enumerable: true,
  get: function get() {
    return _BabelRegister.default;
  }
});
Object.defineProperty(exports, "Generator", {
  enumerable: true,
  get: function get() {
    return _Generator.default;
  }
});
Object.defineProperty(exports, "randomColor", {
  enumerable: true,
  get: function get() {
    return _randomColor.default;
  }
});
Object.defineProperty(exports, "delay", {
  enumerable: true,
  get: function get() {
    return _delay.default;
  }
});
Object.defineProperty(exports, "cleanRequireCache", {
  enumerable: true,
  get: function get() {
    return _cleanRequireCache.default;
  }
});
exports.traverse = exports.parser = exports.t = exports.chokidar = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _address() {
  const data = _interopRequireDefault(require("address"));

  _address = function _address() {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function _chalk() {
    return data;
  };

  return data;
}

function _crossSpawn() {
  const data = _interopRequireDefault(require("cross-spawn"));

  _crossSpawn = function _crossSpawn() {
    return data;
  };

  return data;
}

function chokidar() {
  const data = _interopRequireWildcard(require("chokidar"));

  chokidar = function chokidar() {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "chokidar", {
  enumerable: true,
  get: function get() {
    return chokidar();
  }
});

function _clipboardy() {
  const data = _interopRequireDefault(require("clipboardy"));

  _clipboardy = function _clipboardy() {
    return data;
  };

  return data;
}

function _debug() {
  const data = _interopRequireDefault(require("debug"));

  _debug = function _debug() {
    return data;
  };

  return data;
}

function _deepmerge() {
  const data = _interopRequireDefault(require("deepmerge"));

  _deepmerge = function _deepmerge() {
    return data;
  };

  return data;
}

function _execa() {
  const data = _interopRequireDefault(require("execa"));

  _execa = function _execa() {
    return data;
  };

  return data;
}

function _lodash() {
  const data = _interopRequireDefault(require("lodash"));

  _lodash = function _lodash() {
    return data;
  };

  return data;
}

function _glob() {
  const data = _interopRequireDefault(require("glob"));

  _glob = function _glob() {
    return data;
  };

  return data;
}

function _portfinder() {
  const data = _interopRequireDefault(require("portfinder"));

  _portfinder = function _portfinder() {
    return data;
  };

  return data;
}

function _got() {
  const data = _interopRequireDefault(require("got"));

  _got = function _got() {
    return data;
  };

  return data;
}

function _resolve() {
  const data = _interopRequireDefault(require("resolve"));

  _resolve = function _resolve() {
    return data;
  };

  return data;
}

function _yargs() {
  const data = _interopRequireDefault(require("yargs"));

  _yargs = function _yargs() {
    return data;
  };

  return data;
}

function _mkdirp() {
  const data = _interopRequireDefault(require("mkdirp"));

  _mkdirp = function _mkdirp() {
    return data;
  };

  return data;
}

function _pkgUp() {
  const data = _interopRequireDefault(require("pkg-up"));

  _pkgUp = function _pkgUp() {
    return data;
  };

  return data;
}

function _mustache() {
  const data = _interopRequireDefault(require("mustache"));

  _mustache = function _mustache() {
    return data;
  };

  return data;
}

function _signale() {
  const data = _interopRequireDefault(require("signale"));

  _signale = function _signale() {
    return data;
  };

  return data;
}

function _rimraf() {
  const data = _interopRequireDefault(require("rimraf"));

  _rimraf = function _rimraf() {
    return data;
  };

  return data;
}

function _yargsParser() {
  const data = _interopRequireDefault(require("yargs-parser"));

  _yargsParser = function _yargsParser() {
    return data;
  };

  return data;
}

function t() {
  const data = _interopRequireWildcard(require("@babel/types"));

  t = function t() {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "t", {
  enumerable: true,
  get: function get() {
    return t();
  }
});

function parser() {
  const data = _interopRequireWildcard(require("@babel/parser"));

  parser = function parser() {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "parser", {
  enumerable: true,
  get: function get() {
    return parser();
  }
});

function traverse() {
  const data = _interopRequireWildcard(require("@babel/traverse"));

  traverse = function traverse() {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "traverse", {
  enumerable: true,
  get: function get() {
    return traverse();
  }
});

function _semver() {
  const data = _interopRequireDefault(require("semver"));

  _semver = function _semver() {
    return data;
  };

  return data;
}

var _cheerio = _interopRequireDefault(require("./cheerio/cheerio"));

var _ssr = require("./ssr");

Object.keys(_ssr).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ssr[key];
    }
  });
});

var _routes = require("./routes");

Object.keys(_routes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _routes[key];
    }
  });
});

var _compatESModuleRequire = _interopRequireDefault(require("./compatESModuleRequire/compatESModuleRequire"));

var _mergeConfig = _interopRequireDefault(require("./mergeConfig/mergeConfig"));

var _isLernaPackage = _interopRequireDefault(require("./isLernaPackage/isLernaPackage"));

var _getFile = _interopRequireDefault(require("./getFile/getFile"));

var _winPath = _interopRequireDefault(require("./winPath/winPath"));

var _winEOL = _interopRequireWildcard(require("./winEOL/winEOL"));

var _parseRequireDeps = _interopRequireDefault(require("./parseRequireDeps/parseRequireDeps"));

var _BabelRegister = _interopRequireDefault(require("./BabelRegister/BabelRegister"));

var _Generator = _interopRequireDefault(require("./Generator/Generator"));

var _randomColor = _interopRequireDefault(require("./randomColor/randomColor"));

var _delay = _interopRequireDefault(require("./delay/delay"));

var _cleanRequireCache = _interopRequireDefault(require("./cleanRequireCache/cleanRequireCache"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
