// @umijs/plugin-initial-state/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RELATIVE_EXPORT_PATH = exports.RELATIVE_EXPORT = exports.RELATIVE_MODEL_PATH = exports.RELATIVE_MODEL = exports.MODEL_NAME = exports.DIR_NAME = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DIR_NAME = 'plugin-initial-state';
exports.DIR_NAME = DIR_NAME;
const MODEL_NAME = 'initialState';
exports.MODEL_NAME = MODEL_NAME;
const RELATIVE_MODEL = (0, _path().join)(DIR_NAME, 'models', MODEL_NAME);
exports.RELATIVE_MODEL = RELATIVE_MODEL;
const RELATIVE_MODEL_PATH = `${RELATIVE_MODEL}.ts`;
exports.RELATIVE_MODEL_PATH = RELATIVE_MODEL_PATH;
const RELATIVE_EXPORT = (0, _path().join)(DIR_NAME, 'exports');
exports.RELATIVE_EXPORT = RELATIVE_EXPORT;
const RELATIVE_EXPORT_PATH = `${RELATIVE_EXPORT}.ts`;
exports.RELATIVE_EXPORT_PATH = RELATIVE_EXPORT_PATH;
