// @umijs/preset-ant-design-pro/lib/chainWebpack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function _path() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function getModulePackageName(module) {
  if (!module.context) return null;
  const moduleRelativePath = module.context.split('node_modules').pop() || '';

  const _moduleRelativePath$s = moduleRelativePath.split(_path().default.sep),
        _moduleRelativePath$s2 = _slicedToArray(_moduleRelativePath$s, 1),
        moduleDirName = _moduleRelativePath$s2[0];

  let packageName = moduleDirName; // handle tree shaking

  if (packageName && packageName.match('^_')) {
    // eslint-disable-next-line prefer-destructuring
    packageName = packageName.match(/^_(@?[^@]+)/)[1];
  }

  return packageName;
}

const webpackPlugin = config => {
  // optimize chunks
  config.optimization // share the same chunks across different modules
  .runtimeChunk(false).splitChunks({
    chunks: 'async',
    name: 'vendors',
    maxInitialRequests: Infinity,
    minSize: 0,
    cacheGroups: {
      vendors: {
        test: module => {
          const packageName = getModulePackageName(module) || '';

          if (packageName) {
            return ['bizcharts', 'gg-editor', 'g6', '@antv', 'l7', 'gg-editor-core', 'bizcharts-plugin-slider'].includes(packageName);
          }

          return false;
        },

        name(module) {
          const packageName = getModulePackageName(module);

          if (packageName) {
            if (['bizcharts', '@antv_data-set'].indexOf(packageName) >= 0) {
              return 'viz'; // visualization package
            }
          }

          return 'misc';
        }

      }
    }
  });
  return config;
};

var _default = webpackPlugin;
exports.default = _default;
